### PR TITLE
Install nix-darwin on flake-based systems

### LINKED ISSUE #398: flake: Add apps for {un,}installer?
If the defaultApp of the flake were `darwin-installer`, installation with flakes would be as easy as `nix run github:LnL7/nix-darwin`.

### PR DESCRIPTION
**Pull request by Enzime**
- flake: add packages.darwin-uninstaller
- flake: add packages.darwin-{option,rebuild}
- Allow flaky installation with darwin-rebuild
- readme: update with new flaky instructions
- installer: match flaky installation logic
- doc: store a copy of known files
- etc: fail if we can't add a file
- etc: use .before-nix-darwin instead of .orig

Closes #398 
Closes #399

Tested installing with `darwin-installer` for a channels based system and `darwin-rebuild` for a flake based system on `aarch64-darwin`

### CODE REVIEW

**Review by sielicki — APPROVED** [review]

**Comment by Enzime** [comment]
@domenkozar can you take a look at this PR? The tests are failing due to an unrelated issue (https://github.com/NixOS/nixpkgs/pull/237557) which looks like it'll be a pretty big change that'll require its own PR

**Comment by domenkozar** [comment]
I can't merge PRs that fail CI, so we'll have to address it.

File: .github/workflows/test.yml
Comment (LnL7) [inline .github/workflows/test.yml:126]: Wouldn't it make more sense to expose the the darwin tools package directly instead of building an entire dummy system as the main usecase here is bootstrapping an install?

**Comment by LnL7** [comment]
@domenkozar I think the merge is only blocked by a missing review, while the pr status is red only the stable checks are marked as required. (at least that was my intent with [#641](https://github.com/LnL7/nix-darwin/pull/641))

File: README.md
Comment (Enzime) [reply README.md:16]: So far, the only thing I've moved to regular activation is creating `/run`.

The things I didn't move to regular activation:

- Copying example configuration to `~/.nixpkgs/darwin-configuration.nix`
- Sourcing the static versions of `/etc/bashrc` and `/etc/zshrc` as `programs.bash.enable` defaults to `true` and `programs.zsh.enable = true;` is uncommented in both `simple.nix` and the example `flake.nix`
- Adding `darwin` to `nix-channel`
- Setting `nix.configureBuildUsers = true;` to migrate old style users to new style users, as there's a check that ensures it's set when old style users are present.

File: .github/workflows/test.yml
Comment (Enzime) [reply .github/workflows/test.yml]: Replaced with `nix run .#darwin-rebuild` :+1:

File: README.md
Comment (LnL7) [reply README.md:16]: This used to be more but more modules have been added since the initial version so that's great.

I think the main functional change in that case is the fact that the known hashes for zshrc/bashrc will always need to be correct since there's no installer that edits the existing files anymore. In that case perhaps it's best to also remove those steps from the installer so that both installation paths work the same.

File: pkgs/nix-tools/default.nix
Comment (LnL7) [inline pkgs/nix-tools/default.nix:1]: Instead of passing a whole system to nix-tools it really only needs references to default path locations, this way we don't need a complete system to build the packages.

```suggestion
{ pkgs
, profile ? "/nix/var/nix/profiles/system"
, nixPackage ? "/nix/var/nix/profiles/default"
, defaultPath ? "/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin"
}:

let
  extraPath = lib.makeBinPath [ nixPackage pkgs.coreutils pkgs.jq pkgs.git ];
in

{
  darwin-option = pkgs.writeProgram "darwin-option"
    {
      inherit profile;
      inherit (pkgs.stdenv) shell;
    }
    ./darwin-option.sh;

  darwin-rebuild = pkgs.writeProgram "darwin-rebuild"
    {
      inherit profile;
      inherit (pkgs.stdenv) shell;
      path = "${extraPath}:${defaultPath}";
    }
    ./darwin-rebuild.sh;
}
```

Comment (Enzime) [reply pkgs/nix-tools/default.nix]: Updated :+1:

File: README.md
Comment (Enzime) [reply README.md:16]: I've removed the logic in `darwin-installer` to update `/etc/bashrc` and `/etc/zshrc`, as well as removed the `nix.configureBuildUsers = true;`.

Comment (Enzime) [reply README.md:16]: @LnL7 this PR is ready to review now :smile: 

I've updated the known hashes for `/etc/nix/nix.conf` and `/etc/{bash,zsh}rc`, I've also added a copy of known files as reference for future.

I've also changed the `etc` activation script to fail if any of the files don't match the known hash and need to be replaced manually.

File: modules/nix/default.nix
Comment (emilazy) [inline modules/nix/default.nix:683]: Would be good to add the hashes from https://github.com/DeterminateSystems/nix-installer if none of these match.

Comment (Enzime) [reply modules/nix/default.nix:683]: The issue with adding hashes from Determinate Systems' Nix installer is that they have version numbers in their `nix.conf` so I suspect the hashes will go out of date very quickly, and I'm also not sure how static the contents of `nix.conf` is with their installer.

Comment (emilazy) [reply modules/nix/default.nix:683]: I see, that's a shame :( Maybe we could coordinate with them to get it to be more, uh, deterministic? Not trying to block the PR on this though.

File: flake.nix
Comment (Enzime) [reply flake.nix]: Updated :+1:

**Comment by Enzime** [comment]
@LnL7 @domenkozar can you take a look at this PR when you have a chance?

File: modules/system/flake-overrides.nix
Comment (LnL7) [inline modules/system/flake-overrides.nix:15]: Won't this make specifying `nix.nixPath` mandatory for flake configurations?
I think wether a system uses `<nixpkgs>` as a channel or not is completely separate from if the configuration itself happens to be built using flakes.

**Review by LnL7 — DISMISSED** [review]
A few small comments, but looks good otherwise.

File: modules/system/flake-overrides.nix
Comment (Enzime) [reply modules/system/flake-overrides.nix:15]: I'll remove this from this PR, the reason why I added it in the first place was primarily because `NIX_PATH` contained `darwin-configuration=` which was not relevant when using flakes. I think it would be good in future to have some more granular options for controlling the values added to `nix.nixPath`.

File: .github/workflows/test.yml
Comment (Enzime) [reply .github/workflows/test.yml:57]: The reason it's necessary in CI is because `nix.conf` contains the GitHub access token

File: modules/system/flake-overrides.nix
Comment (Enzime) [reply modules/system/flake-overrides.nix]: Removed :+1:

Comment (LnL7) [reply modules/system/flake-overrides.nix]: Ah, luckily nix.nixPath can also be an attribute set so I think it shouldn't be that hard to conditionally add the configuration path.

File: .github/workflows/test.yml
Comment (LnL7) [reply .github/workflows/test.yml:57]: Makes the testing setup a bit messy, but the way those currently run isn't ideal anyway. With a nicer setup a lot more things during activation could be validated properly.

**Review by LnL7 — APPROVED** [review]

File: .github/workflows/test.yml
Comment (Enzime) [reply .github/workflows/test.yml:57]: I'm happy to investigate this further in another PR, any things that come to mind that we should test for?

File: README.md
Comment (emilazy) [inline README.md:83]: Please collapse this into a single line; it renders as a `<p>` on GitHub and adds a weird line break.

File: README.md
Comment (emilazy) [inline README.md:102]: (as above)

File: README.md
Comment (emilazy) [inline README.md:93]: This should reference `nixpkgs.hostPlatform` now.

File: README.md
Comment (emilazy) [inline README.md:135]: These make sense if the reader expanded the migration section, but not for the from-scratch one. I suggest adding wording to the from-scratch section telling the reader to replace `simple` with their system's hostname, as is best practice, and omitting the `#...` here.

File: README.md
Comment (emilazy) [inline README.md:138]: Maybe say `hostname -s` instead:

```console
emily@yuyuko ~> hostname 
yuyuko.emily.moe
emily@yuyuko ~> hostname -s
yuyuko
```

The latter is used by `darwin-rebuild`.

File: flake.nix
Comment (emilazy) [inline flake.nix:42]: The uninstaller should be updated to also remove the `run` line from `synthetic.conf` and poke APFS appropriately.

It also completely removed `/etc/nix/nix.conf`, which is not great. It should restore the pre-nix-darwin backup file, if present.

It also printed an `rm` error when I ran it on my test VM:

```console
…
removing service org.nixos.activate-system
removing service org.nixos.nix-daemon
configuring networking...
rm: cannot remove '/Library/LaunchDaemons/org.nixos.nix-daemon.plist': No such file or directory
grep: /etc/bashrc: No such file or directory
Found no nix-daemon.sh reference in /etc/bashrc
add this snippet back to /etc/bashrc:
…
```

Maybe that's expected though.

**Review by emilazy — CHANGES_REQUESTED** [review]
Thank you for working on this! I've left review comments for various nits; the other things I'd like to see before merging are:

* The hash for the `nix.conf` produced by the latest version of the Determinate Systems installer with default settings to be added, as the error for an unknown `nix.conf` will be confusing for new years and leaves the system in a half-activated state. The SHA-256 is `f3e03d851c240c1aa7daccd144ee929f0f5971982424c868c434eb6030e961d4` and the contents are:

  ```
  # Generated by https://github.com/DeterminateSystems/nix-installer, version 0.10.0.
  experimental-features = nix-command flakes auto-allocate-uids
  build-users-group = nixbld
  auto-optimise-store = true
  extra-nix-path = nixpkgs=flake:nixpkgs
  bash-prompt-prefix = (nix:$name)\040
  ```

  (plus trailing newline)

* Instructions for uninstalling a flakes-based system in `README.md`, whether in the uninstallation section (preferable as I think that's where people are most likely to look first) or the flakes section.

**Comment by emilazy** [comment]
Thinking about https://github.com/LnL7/nix-darwin/pull/687/commits/77936c4e541d12b99bbddf088ff2ddeeb95be2c1, it might be good to move the hash checks into the `checks` activation script so that mismatches don't cause partially-activated configurations, which is the thing it's trying to avoid in the first place and which I ran into when testing this.

File: README.md
Comment (LnL7) [reply README.md:138]: FYI there's is #676 which would change the behaviour slightly.

**Comment by LnL7** [comment]
That's a good idea, if we want to change a bunch of etc behaviour might be easier to do that in a separate PR however.

File: README.md
Comment (emilazy) [reply README.md:138]: Yeah, I'm agnostic to which we use; I might slightly prefer the current solution as I've had Bonjour clashes before. Mostly just want to document the correct one whichever it is.

File: README.md
Comment (Enzime) [reply README.md]: Updated :+1:

File: README.md
Comment (Enzime) [reply README.md]: Updated :+1:

File: README.md
Comment (Enzime) [reply README.md]: Updated :+1:

File: README.md
Comment (Enzime) [reply README.md]: Updated :+1:

File: flake.nix
Comment (Enzime) [reply flake.nix:42]: I think the `darwin-uninstaller` needs extra work and my preference would be to focus on that in another PR

File: modules/nix/default.nix
Comment (Enzime) [reply modules/nix/default.nix]: Added `nix.conf` from the latest DeterminateSystems nix-installer :+1:

**Comment by Enzime** [comment]
I've incorporated some of @emilazy's suggestions as well as adding a new commit just to use the suffix `before-nix-darwin` instead of `orig` as it makes the source of the file clearer

File: modules/services/tailscale.nix
Comment (emilazy) [inline modules/services/tailscale.nix:80]: Should be `.before-nix-darwin`, this will presumably break Tailscale.

**Review by emilazy — CHANGES_REQUESTED** [review]
Alright, I think this is good to go. The installer and `nix.conf` issues are a pain but we can solve them separately and I don't want to hold this up any further. Just one more thing though...

File: modules/services/tailscale.nix
Comment (Enzime) [reply modules/services/tailscale.nix]: Good catch, fixed :+1:

**Review by emilazy — APPROVED** [review]
Let's ship it! I'll open an issue for the uninstaller stuff but this looks like a great improvement to me. Thank you for all your work on this.

**Pull request merged**
